### ISSUE TITLE
Differences in Model Weight Behavior Between Keras 2.15 and Keras 3

### ISSUE THREAD

**Issue by edwardyehuang** [keras-team/keras#21676]
I attempted to migrate name-based `Keras 2.x` weights in `.h5` format to `Keras 3` by loading the `.h5` weights into the model and subsequently saving them as either `.weights.h5` or `.keras` files using `model.save()`. 

The resulting weight files were successfully loaded and functioned correctly in `Keras 2.15`. 

However, upon transitioning to `Keras 3`, these weights failed to load or operate as expected.

I am currently investigating this issue by analyzing the source code. If you have any useful suggestions, I would appreciate your input.

**Comment by dhantule** [comment]
Hi @edwardyehuang, thanks for reporting this. 
Could you please refer this issue #20081 and let me know if it helps?

**Comment by innat** [comment]
https://github.com/keras-team/keras/issues/19060

**Comment by edwardyehuang** [comment]
@dhantule  This may be related to the trackable weights. See my latest comments in #21078

**Comment by dhantule** [comment]
@edwardyehuang, we'll look into this and update you.

**Comment by edwardyehuang** [comment]
It appears that implementing a straightforward name-based weight loading mechanism, without batch loading, is necessary to ensure compatibility of .h5 weight files between Keras 2.x and Keras 3. I will handle that

**Comment by monicadsong** [comment]
Hi @edwardyehuang -

The Keras team would like to learn more about your Keras usage. Please thumbs up this comment if you are open to speaking with us and we'll reach out to you over email.

Thank you so much!
Monica, Keras Product Manager

**Comment by edwardyehuang** [comment]
> Hi [@edwardyehuang](https://github.com/edwardyehuang) -
> 
> The Keras team would like to learn more about your Keras usage. Please thumbs up this comment if you are open to speaking with us and we'll reach out to you over email.
> 
> Thank you so much! Monica, Keras Product Manager

Yes, it is an honor to have the opportunity to engage directly with the Keras team in efforts to further enhance Keras.

**Comment by georgeneedles60-hub** [comment]
thank you

**Comment by dhantule** [comment]
Hi @edwardyehuang, can you provide some reproducible code to reproduce the issue ?

**Comment by edwardyehuang** [comment]
https://github.com/edwardyehuang/iSeg/blob/master/saver/h5_saver.py

The above is my implementation of name-based weight loading, which assumes that the order of the h5 weights and symbolic_weights differs. It compares the similarity of names between the h5 weights and symbolic_weights.

**Comment by edwardyehuang** [comment]
I am now able to perform inference on all my previous models trained with TensorFlow 2.15 and Keras 2.15 using the Keras 3.10 environment, yielding correct inference results. However, training remains unsuccessful, producing anomalous outputs (e.g. loss) even in the first step, and the issue is currently under investigation.

Currently found:

- Mixed precision training exhibits incorrect behavior when compared to Keras 2.15.

**Comment by divyashreepathihalli** [comment]
@edwardyehuang - can you please provide a repro for this issue? It would help us debug better.

Here is what works right now

for full model migration 
```
model = keras.models.load_model("keras2_model.h5")

# Save in Keras 3 formats
model.save("keras3_model.keras")  # New format
model.save_weights("keras3_weights.weights.h5")
```

for weights only migration 
```
# Recreate model architecture
model = create_identical_architecture()
model(dummy_input)  # Build the model

# Load legacy weights
model.load_weights("keras2_weights.h5", by_name=True, skip_mismatch=True)

# Save in Keras 3 format
model.save_weights("keras3_weights.weights.h5")
```


Let me know if this helps. If not, a repro would help debug better.

**Comment by edwardyehuang** [comment]
> This issue is stale because it has been open for 14 days with no activity. It will be closed if no further activity occurs. Thank you.

Will update soon
